### ISSUE TITLE
improvement in apps/s_cb.c:security_callback for op=SSL_SECOP_OTHER_SIGALG

### ISSUE THREAD

**Issue by lzsiga** [openssl/openssl#8385]
Hi I'd like to suggest a little improvement regarding `security_callback`
`apps`/`s_cb.c`/`security_callback_debug` for the case when
(op&SSL_SECOP_OTHER_TYPE)==SSL_SECOP_OTHER_SIGALG

This callbacks evaluates byte `salg[1]` (which is the LSB of `sigalg`
from `t1_lib.c:sigalg_lookup_tbl[i]`), to pick one of `TLSEXT_signature_anonymous`/`_rsa`/`_dsa`/`_ecdsa`.

But some `sigalg` values cannot be decoded this way, for example `TLSEXT_SIGALG_rsa_pss_rsae_sha256`=0x0804, `TLSEXT_SIGALG_rsa_pss_pss_sha256`=0x0809; their LSBs being 0x04 and 0x09.

I would like to suggest to also pass the value of `sig` from `sigalg_lookup_tbl[i]`,
so that the callback could call its `get_sigtype` function.

Patch for ssl/t1_lib.c:
```
--- t1_lib.cold 2019-02-26 15:15:30.000000000 +0100
+++ t1_lib.c    2019-03-01 17:24:24.804282277 +0100
@@ -1557,7 +1557,7 @@
 /* Check to see if a signature algorithm is allowed */
 static int tls12_sigalg_allowed(SSL *s, int op, const SIGALG_LOOKUP *lu)
 {
-    unsigned char sigalgstr[2];
+    unsigned char sigalgstr[4];
     int secbits;
 
     /* See if sigalgs is recognised and if hash is enabled */
@@ -1623,6 +1623,8 @@
     /* Finally see if security callback allows it */
     sigalgstr[0] = (lu->sigalg >> 8) & 0xff;
     sigalgstr[1] = lu->sigalg & 0xff;
+    sigalgstr[2] = (lu->sig >> 8) & 0xff;
+    sigalgstr[3] = lu->sig & 0xff;
     return ssl_security(s, op, secbits, lu->hash, (void *)sigalgstr);
```
Patch for apps/s_cb.c:
```
--- s_cb.cold	2019-02-26 15:15:30.000000000 +0100
+++ s_cb.c	2019-03-01 17:31:51.216270054 +0100
@@ -1373,26 +1373,15 @@
         {
             const unsigned char *salg = other;
             const char *sname = NULL;
-            switch (salg[1]) {
-            case TLSEXT_signature_anonymous:
-                sname = "anonymous";
-                break;
-            case TLSEXT_signature_rsa:
-                sname = "RSA";
-                break;
-            case TLSEXT_signature_dsa:
-                sname = "DSA";
-                break;
-            case TLSEXT_signature_ecdsa:
-                sname = "ECDSA";
-                break;
-            }
+            int lu_sigalg= (salg[0]<<8) + salg[1];
+            int lu_sig=    (salg[2]<<8) + salg[3];
+            sname= get_sigtype (lu_sig);
 
             BIO_puts(sdb->out, OBJ_nid2sn(nid));
             if (sname)
                 BIO_printf(sdb->out, ", algorithm=%s", sname);
             else
-                BIO_printf(sdb->out, ", algid=%d", salg[1]);
+                BIO_printf(sdb->out, ", algid=%d(raw=0x%x)", lu_sig, lu_sigalg);
             break;
         }
```
Result in 's_server -security_debug_verbose'
before:
```
 Security callback: Shared Signature Algorithm digest=SHA256, algid=9,  security bits=128: yes
 Security callback: Shared Signature Algorithm digest=SHA384, algid=10, security bits=192: yes
 Security callback: Shared Signature Algorithm digest=SHA512, algid=11, security bits=256: yes
```
after:
```
 Security callback: Shared Signature Algorithm digest=SHA256, algorithm=RSA-PSS, security bits=128: yes
 Security callback: Shared Signature Algorithm digest=SHA384, algorithm=RSA-PSS, security bits=192: yes
 Security callback: Shared Signature Algorithm digest=SHA512, algorithm=RSA-PSS, security bits=256: yes
```

**Comment by lzsiga** [comment]
Sorry, forgot to say, it's for version 1.1.1b

**Comment by mattcaswell** [comment]
IMO this is not the right way to go. It seems quite ugly to me to tack additional bytes on to the sigalg identifier in the security callback. This identifier matches exactly with the identifier as defined in the RFCs and to suddenly make it into a 4 byte structure with additional meaning seems quite ugly.

Really I think the security_callback_debug callback in s_cb.c should have its own lookup table of all known sigalgs and work from that instead. I'd be happy to see PRs that do this.

**Comment by lzsiga** [comment]
Hi, mattcaswell, thank you for your reply.
I did try something regarding s_cb.c, now it generates output like this:
```
Security callback: Shared Signature Algorithm digest=SHA224, algorithm=DSA, security bits=112: yes
Security callback: Shared Signature Algorithm digest=SHA1, algorithm=DSA, security bits=80: yes
Security callback: Shared Signature Algorithm digest=SHA256, algorithm=DSA, security bits=128: yes
Security callback: Shared Signature Algorithm digest=SHA384, algorithm=DSA, security bits=192: yes
Security callback: Shared Signature Algorithm digest=SHA512, algorithm=DSA, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp256r1_sha256, security bits=128: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp384r1_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=ecdsa_secp521r1_sha512, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=rsa_pss_pss_sha256, security bits=128: yes
Security callback: Shared Signature Algorithm scheme=rsa_pss_pss_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=rsa_pss_pss_sha512, security bits=256: yes
Security callback: Shared Signature Algorithm scheme=rsa_pss_rsae_sha256, security bits=128: yes
Security callback: Shared Signature Algorithm scheme=rsa_pss_rsae_sha384, security bits=192: yes
Security callback: Shared Signature Algorithm scheme=rsa_pss_rsae_sha512, security bits=256: yes
```
So it writes 'scheme=_scheme_' for the schemes listed in rfc8446, for the others it writes 'digest=_digest_, method=_method_'. If neither worked, it would write this:
```
Security callback: Shared Signature Algorithm scheme=unknown(0x0e01), security bits=128: yes
Security callback: Shared Signature Algorithm scheme=unknown(0x0e02), security bits=192: yes
Security callback: Shared Signature Algorithm scheme=unknown(0x0e03), security bits=256: yes
```
Attaching the patch, please have a look at it.
[s_cb.diff.txt](https://github.com/openssl/openssl/files/2936545/s_cb.diff.txt)
lzsiga (Lorinczy, Zsigmond)

**Comment by mattcaswell** [comment]
Looks fairly reasonable but please raise this as a pull request so that it can be reviewed properly

**Comment by lzsiga** [comment]
Ok, created pull request #8429
I forgot to mention, but I'd like to tag it as 'trivial change'

**Comment by lzsiga** [comment]
Well, I closed #8249 instead created #8445 -- same content, with corrections from mattcaswell

**Comment by lzsiga** [comment]
Seems to be solved:)
Thank you guys.
